### ISSUE TITLE
Yahoo Finance returns 502 Bad Gateway

### ISSUE THREAD

**Issue by joshuaulrich** [joshuaulrich/quantmod#157]
After switching from HTTP to HTTPS (#149), it looks like Yahoo has now changed the URL.  

```r
R> getSymbols("AAPL", src="yahoo")
Error in download.file(paste(yahoo.URL, "s=", Symbols.name, "&a=", from.m,  : 
  cannot open URL 'https://ichart.finance.yahoo.com/table.csv?s=AAPL&a=0&b=01&c=2007&d=4&e=16&f=2017&g=d&q=q&y=0&z=AAPL&x=.csv'
In addition: Warning message:
In download.file(paste(yahoo.URL, "s=", Symbols.name, "&a=", from.m,  :
  cannot open URL 'https://ichart.finance.yahoo.com/table.csv?s=AAPL&a=0&b=01&c=2007&d=4&e=16&f=2017&g=d&q=q&y=0&z=AAPL&x=.csv': HTTP status was '502 Bad Gateway'
```

The "Download Data" link on the [historical data page for AAPL](https://finance.yahoo.com/quote/AAPL/history?p=AAPL) is now:
https://query1.finance.yahoo.com/v7/finance/download/AAPL?period1=1492379145&period2=1494971145&interval=1d&events=history&crumb=sgdBTt48gg.

**Comment by arbitamrazian** [comment]
I also noticed this bug today, but how do we deal with the crumb parameter?  It's required since after  [https://query1.finance.yahoo.com/v7/finance/download/AAPL?period1=1492379145&period2=1494971145&interval=1d&events=history](url) (i.e. without the crumb param) I get 
`{
    "finance": {
        "error": {
            "code": "Unauthorized",
            "description": "Invalid cookie"
        }
    }
}`

Any ideas here?

**Comment by joshuaulrich** [comment]
You need to establish a session with the Yahoo Finance server in order to get the crumb. That will probably require using curl (via either the RCurl or curl R package).

**Comment by machuck** [comment]
Can someone post an example of how to use RCurl + getSymbols etc?

**Comment by lompoc42** [comment]
I'm sorry, but I'm confused as to how to use the updates. How do I update the functions with the changes you just made?

**Comment by joshuaulrich** [comment]
@lompoc42 one of these:

```r
remotes::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
# or
devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
```

Or wait until it's merged to master and use:
```r
remotes::install_github("joshuaulrich/quantmod")
# or
devtools::install_github("joshuaulrich/quantmod")
```

**Comment by lompoc42** [comment]
THANK YOU SO MUCH!!!!!!!!!!!

**Comment by lompoc42** [comment]
Getting an error with the update getSymbols.yahoo

Error in `/.default`(fr[, 4], fr[, 6]) : 
  non-numeric argument to binary operator

**Comment by joshuaulrich** [comment]
@lompoc42 Please provide the command that causes that error.  It's nearly impossible to help without it.

**Comment by lompoc42** [comment]
In RStudio
```
>stock = new.env()
>getSymbols.yahoo(Symbols = '^GSPC',env = stock)
```

**Comment by yankodev** [comment]
Hey Joshua, thanks for giving us quantmod, and thanks for the quick fix.

After installing the update I can download FB and APPL historical prices with, e.g., loadSymbols("FB").

I get the same error as lompoc42 when trying ^GSPC, ^GDAXI, SIE.DE, or LQD. It seems the query string varies with the ticker (just a shot in the dark, no time to check).

Cheers,
Yanko

**Comment by lompoc42** [comment]
Yeah, it works for some but not for others

**Comment by joshuaulrich** [comment]
The issue with the symbols listed should now be resolved. Thanks for the testing and feedback!

**Comment by lompoc42** [comment]
The '502 Bad Gateway' error is back. I didn't change anything which was working before and now no longer working.

**Comment by lompoc42** [comment]
I will answer my own question:

Had to uninstall and re-install quantmod then it worked

**Comment by lompoc42** [comment]
Not sure what's happened, but it's not working anymore

**Comment by bdfden** [comment]
It is now working for me.  Thank you for your help.  However, it looks like it is only returning data going back to 2014.  Also, the finance.yahoo.com site seems to be having issues returning older data.

**Comment by joshuaulrich** [comment]
> Not sure what's happened, but it's not working anymore

@lompoc42 it works fine for me from a fresh install on another machine. I assume you're running from RStudio; have you tried restarting RStudio?  Unload/reload doesn't always work well.  If that doesn't work, see if it works if you run from the GUI that comes with R and/or the terminal.

> ...it looks like it is only returning data going back to 2014. Also, the finance.yahoo.com site seems to be having issues returning older data.

@bdfden I just pulled data from 2007.  If you're only getting data back to 2014 for a specific symbol, and you get the same behavior for the same instrument on the website, then there's no issue with quantmod.

**Comment by bdfden** [comment]
Agreed - this issue seems to be with Yahoo.  Thanks again.

**Comment by wave-electron** [comment]
@joshuaulrich 

Just want to thank you for the great work on the quantmod package. I spent a few weeks developing a shiny app that could mimic yahoo charts in April 2017...my goal was to showcase what is now possible using the shiny framework & R's immense library of packages. I'm using the highcharts R package which I also highly recommend for charting. 

Here is a screen shot. 

![screen shot 2017-05-18 at 9 37 03 am](https://cloud.githubusercontent.com/assets/16273780/26180637/f9cdb24a-3bad-11e7-9173-ab1bc4467c8b.png)

When I noticed the 502 Bad Gateway yesterday I feared the worse. But I've implemented your update and the app is working as per usual. 

So thanks very much for all your great work!

**Comment by ellisvalentiner** [comment]
Woah. Your shiny app looks much better than mine!

**Comment by mcohen01** [comment]
no big deal if you don't have time to support this, i can wait, but here's the error i'm getting trying to install from the branch:

```
> devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
Downloading GitHub repo joshuaulrich/quantmod@157_yahoo_502
from URL https://api.github.com/repos/joshuaulrich/quantmod/zipball/157_yahoo_502
Installing quantmod
trying URL 'https://cran.rstudio.com/bin/macosx/mavericks/contrib/3.3/zoo_1.8-0.tgz'
Content type 'application/x-gzip' length 887574 bytes (866 KB)
==================================================
downloaded 866 KB

Installing zoo
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file  \
  --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  '/private/var/folders/64/7v2m4_bd57z9kt_dttsc79zhrhzc_h/T/RtmpRadWAr/devtools5bc1187d120b/zoo'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library'  \
  --install-tests 

* installing *binary* package ‘zoo’ ...
* DONE (zoo)
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file  \
  --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  '/private/var/folders/64/7v2m4_bd57z9kt_dttsc79zhrhzc_h/T/RtmpRadWAr/devtools5bc14faece28/joshuaulrich-quantmod-7e7744f'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library'  \
  --install-tests 

* installing *source* package ‘quantmod’ ...
** libs
gfortran-4.8   -fPIC  -g -O2  -c ohlcq.f -o ohlcq.o
make: gfortran-4.8: No such file or directory
make: *** [ohlcq.o] Error 1
ERROR: compilation failed for package ‘quantmod’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
Error: Command failed (1)
```

```
> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X El Capitan 10.11.6

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
[1] httr_1.2.1      R6_2.2.0        tools_3.3.2     withr_1.0.2     curl_2.3       
[6] memoise_1.0.0   git2r_0.18.0    digest_0.6.12   devtools_1.12.0
```

**Comment by joshuaulrich** [comment]
@mcohen01 See the second-to-last bullet in the Installation section of quantmod's README:
https://github.com/joshuaulrich/quantmod#installation

**Comment by jzerp** [comment]
Thanks for Quantmod Joshua, I'm very new to R and I have been using Quantmod for a couple of months and its been very useful. However, I ran into the same  problem other users have talked about. Just wanted to ask you if you are planning to release a new version or should I stick to the fixes you listed before. I tried to run them but I'm getting errors. Thanks again!

**Comment by richjung** [comment]
Thank you for your help. Should I make any change in Ad(get(symbol[i]]) to get prices from multiple tickers?

**Comment by Gabylin** [comment]
Thank you so much, it works now after installed devtools provided by Joshua. I can check stock data as usual now.

**Comment by bdfden** [comment]
Seems like this might be a bigger issue with Yahoo.  I have the update working, but my historical numbers have totally changed.  I think this might be due to Yahoo dropping info on splits/dividends.   See https://yahoo.uservoice.com/forums/382977-finance/suggestions/19310356-https-ichart-finance-yahoo-com-table-csv-s-spy-a#comments for further discussion.

**Comment by aguadavid** [comment]
Sorry if this is a dumb question, but how to we deploy this update to shinyapps.io?  I have it working locally, but even though I republished, I'm still having this problem on shinyapps.io... thanks!

**Comment by ianarvin** [comment]
@joshuaulrich Thank you so much for all you do for the R community. I tried to add your quantmod fix for getting Yahoo stock prices and received this error. How do I fix? Thanks!

> remotes::install_github("joshuaulrich/quantmod")
Downloading GitHub repo joshuaulrich/quantmod@master
Error in if (file.exists(dest) && file.mtime(dest) > file.mtime(lib) &&  : 
  missing value where TRUE/FALSE needed
In addition: Warning message:
In missing_devel_warning(pkgdir) :
  Package quantmod has compiled code, but no suitable compiler(s) were found. Installation will likely fail.
  Install Rtools and make sure it is in the PATH.
Error in if (file.exists(dest) && file.mtime(dest) > file.mtime(lib) &&  : 
  missing value where TRUE/FALSE needed

**Comment by Javdat** [comment]
`getQuote` has probably related issue:

```
getQuote("AAPL")

 Error in open.connection(con, "rb") : HTTP error 404. 
```

The link has changed, but I am not quite sure, what is the new one, yet.

**Comment by joshuaulrich** [comment]
@ianarvin See the second-to-last bullet in the [Installation section](https://github.com/joshuaulrich/quantmod#installation) of quantmod's README.

@Javdat Yeah, I expected something like that to happen... but didn't want to change anything until it did.  That command still works for me, but Yahoo may be in the process of migrating servers and I'm still hitting one of the old ones.

@aguadavid This package's issue tracker isn't the forum for troubleshooting shinyapp issues.  I don't mean to be rude, but off-topic comments make it harder for me to track comments for this issue.

@bdfden Thanks for the report!  I noted that the OHL columns were now adjusted, and that made me wonder about the correctness of the new data.  I would really appreciate if you (or anyone else) could do some cross-checking and/or validation!

@jzerp You should keep using the fixes on this branch.  I will push an update to CRAN shortly, but I would first like to ensure the fixes are reasonably correct.  I do not want to burden CRAN with multiple releases in a very short period of time.

**Comment by lompoc42** [comment]
@bdfden Which tickers are you looking at? I just checked multiple tickers from Yahoo's adjusted close to WIKI and Quotemedia data; all checked out fine.

**Comment by Javdat** [comment]
Here is a quick test (I am not sure if it should go here or get its own issue).

```
appl.yahoo <- getSymbols("AAPL", from = "2001-01-01", auto.assign = FALSE)
appl.google <- getSymbols("AAPL", from = "2001-01-01", auto.assign = FALSE, src = "google")
appl.quandl <- Quandl("EOD/AAPL", api_key="xxxxxxxxxxxxxxxxx", start_date="2001-01-01", type = "xts")

# Before split 09-06-2014
appl.yahoo["2014-06-02"]
appl.google["2014-06-02"]
appl.quandl["2014-06-02"]
```

Out:
```
> appl.yahoo["2014-06-02"]
           AAPL.Open AAPL.High AAPL.Low AAPL.Close AAPL.Volume AAPL.Adjusted
2014-06-02    633.96    634.83    622.5     628.65    92337700      89.80714
> appl.google["2014-06-02"]
           AAPL.Open AAPL.High AAPL.Low AAPL.Close AAPL.Volume
2014-06-02     90.57     90.69    88.93      89.81    92337903
> appl.quandl["2014-06-02"]
             Open   High   Low  Close   Volume Dividend Split Adj_Open Adj_High  Adj_Low Adj_Close Adj_Volume
2014-06-02 633.96 634.83 622.5 628.65 13191100        0     1 85.58922 85.70667 84.04203  84.87233   92337700
```
Splits are reflected in yahoo Adjusted, but not in OHLC.

```
# After split, before divdend at 2017-05-11 (ex-diff)
appl.yahoo["2017-05-10::2017-05-12"]
appl.google["2017-05-10::2017-05-12"]
appl.quandl["2017-05-10::2017-05-12"]
```

Out:
```
> appl.yahoo["2017-05-10::2017-05-12"]
           AAPL.Open AAPL.High AAPL.Low AAPL.Close AAPL.Volume AAPL.Adjusted
2017-05-10    153.63    153.94   152.11     153.26    25805700        153.26
2017-05-11    152.45    154.07   152.31     153.95    27255100        153.95
2017-05-12    154.70    156.42   154.67     156.10    32527000        156.10
> appl.google["2017-05-10::2017-05-12"]
           AAPL.Open AAPL.High AAPL.Low AAPL.Close AAPL.Volume
2017-05-10    153.63    153.94   152.11     153.26    25805692
2017-05-11    152.45    154.07   152.31     153.95    27255058
2017-05-12    154.70    156.42   154.67     156.10    32527017
> appl.quandl["2017-05-10::2017-05-12"]
             Open   High    Low  Close   Volume Dividend Split Adj_Open Adj_High  Adj_Low Adj_Close Adj_Volume
2017-05-10 153.63 153.94 152.11 153.26 25805692     0.00     1 153.0039 153.3126 151.4901  152.6354   25805692
2017-05-11 152.45 154.07 152.31 153.95 27255058     0.63     1 152.4500 154.0700 152.3100  153.9500   27255058
2017-05-12 154.70 156.42 154.67 156.10 32527017     0.00     1 154.7000 156.4200 154.6700  156.1000   32527017
> 
```
Dividends are not reflected on OHLC or Adjusted.

**Comment by bdfden** [comment]
Check out the historical data for SPXL.   They did a 4:1 split on May 1, but the adjusted closing prices seem to be out of whack. 

Also, I've been using the library for years to track a system that trades the SPY.   I can't quite figure it out, but those results have totally changed, and they only thing I could come up with was a potential problem with the dividend.

**Comment by ianarvin** [comment]
@joshuaulrich Where do I find the installation README for quantmod? I couldn't find your reference to second to last bullet point.

**Comment by lompoc42** [comment]
@bdfden You're right. I had previously checked out non-divided items I guess! Everything with a dividend is off... Wonderful...

**Comment by joshuaulrich** [comment]
@ianarvin I included a link to it in my previous comment... https://github.com/joshuaulrich/quantmod#installation

@Javdat Thanks for the information! Yahoo previously provided raw prices and a split- and dividend-adjusted close. Google provided split-adjusted prices. It's good to know that the Yahoo adjusted close is no longer dividend-adjusted.

@bdfden, @lompoc42 you can use `adjustOHLC()` to adjust the raw prices for splits and dividends.

**Comment by bojanbis** [comment]
Perhaps I'm installing the wrong version of quantmod, but I'm still getting the error that the link is incorret:

getSymbols("AAPL", from = "2001-01-01", auto.assign = FALSE)
Error in download.file(paste(yahoo.URL, "s=", Symbols.name, "&a=", from.m,  :
  cannot open URL 'http://ichart.finance.yahoo.com/table.csv?s=AAPL&a=0&b=01&c=2001&d=4&e=18&f=2017&g=d&q=q&y=0&z=AAPL&x=.csv'


This link seems to work (replace "IBM" with ticker of y our choice); it returns some JSON code that needs to be parsed...

https://query2.finance.yahoo.com/v8/finance/chart/IBM?formatted=true&lang=en-US&region=US&period1=1487372400&period2=1495058400&interval=1d&events=div%7Csplit&corsDomain=finance.yahoo.com

**Comment by bdfden** [comment]
Yes - I've always used adjustOHLC() as part of my code.  In addition, I tried both use.adjusted = T and F.  Both T and F return the exact same results, and they are both lower than what I've seen historically.   Again, dividends might not be the issue, but it's the only thing I can think of.

**Comment by lompoc42** [comment]
@bdfden @joshuaulrich 

I'm getting an error when trying to use adjustOHLC()

```
>stock = new.env()
>getSymbols(Symbols = 'TLT', src = 'yahoo', from = '1910-01-01', env = stock)
>adjustOHLC('TLT')
```
yields

```
+ Error in read.table(file = file, header = header, sep = sep, quote = quote,  : 
+ no lines available in input
```
I get the same error if I try using stock$TLT instead of 'TLT'. Maybe this is why this process hasn't worked for you @bdfden ?

**Comment by bdfden** [comment]
@lompoc42   I'm now getting this error when I try use.Adjusted=F.  Looks similar to the one you posted.

> tail(adjustOHLC(symbols$SPXL, use.Adjusted=T))
           SPXL.Open SPXL.High SPXL.Low SPXL.Close SPXL.Volume SPXL.Adjusted
2017-05-10     32.92     33.11    32.82      33.11      699300         33.11
2017-05-11     32.91     32.98    32.40      32.93     1253200         32.93
2017-05-12     32.78     32.83    32.60      32.76      723000         32.76
2017-05-15     32.92     33.32    32.91      33.22     1409300         33.22
2017-05-16     33.41     33.42    32.99      33.16     1256900         33.16
2017-05-17     32.33     32.57    31.38      31.44     3005800         31.44
> tail(adjustOHLC(symbols$SPXL, use.Adjusted=F))
Error in read.table(file = file, header = header, sep = sep, quote = quote,  : 
  no lines available in input

**Comment by herrold** [comment]
@joshuaulrich 

Before investing too much time, this post from yesterday in the Yahoo support form may be of interest:

https://forums.yahoo.net/t5/Yahoo-Finance-help/Is-Yahoo-Finance-API-broken/td-p/250503/page/3

Official 'Hoo Staff Nixon
Official 'Hoo Staff

yesterday

Re: Is Yahoo! Finance API broken?

Hi All - This feature was discontinued by the Finance team and they will not be reintroducing that functionality. 

The Yahoo Finance Feedback Forum is the place where you can make product suggestions and provide feedback. We’re always trying to improve our products and use your feedback to inform changes.  Here’s the url: https://yahoo.uservoice.com/forums/382977

**Comment by lompoc42** [comment]
@bdfden @joshuaulrich 

Got it to work!!!! If you explicitly use the symbol.name argument inside of adjustOHLC() it works!! Have checked several items now

**Comment by bdfden** [comment]
@lompoc42   Sorry - can you post an example?  I can't get it to work by trying adjustOHLC("SPY",use.Adjusted=F).  Thanks!

**Comment by lompoc42** [comment]
@bdfden try:

I'm having to provide adjustOHLC the path to my data directly and specifying the symbol.name argument

`adjustOHLC(SPY.data, use.Adjusted=F, symbol.name='SPY')`

**Comment by lompoc42** [comment]
Suddenly stopped working

**Comment by lompoc42** [comment]
Finding it doesn't work for everything, unfortunately. VNQ and RWO did not work, and I'm sure there are others, but it works most of the time. I put a try exemption in for running it over a basket and it only kicked back maybe 4 or 5 errors

**Comment by bdfden** [comment]
Gotcha.  I tried running it for SPXL.  I was able to get it to work for both use.Adjusted=T&F, but the returned data all matched the original SPXL dataset which causes a major error in the model (shows giant loss after the split).   

I'm thinking I might have to check out Quandl...

**Comment by misantroop** [comment]
While I'm not entirely familiar with the internals of Quantmod, it's worth noting that the adjustments for splits are done sporadically. The "adjusted close" field is completely unusable as the formula Yahoo now uses is erroneous. The other prices aren't typically adjusted for splits either - for an example, take a look at BLL which had a split recently.

**Comment by rogerjbos** [comment]
This bug is MINOR so I will post it here instead of opening an issue:
 
I was an example from someone else and got an error:
> getSymbols("IBM", from="1900-01-01")
NAs introduced by coercion to integer rangeError in curl::curl_download(paste0(yahoo.URL, Symbols.name, "?period1=",  : 
  HTTP error 400.

This seems to happen with dates before the origin, so an easy fix would be to wrap max() statement around line 274 as shown below:

    from.posix <- max(0, as.integer(as.POSIXct(as.Date(from, origin = "1970-01-01"))))

I know we should rely on people using a reasonable date range, but no everyone will and the error is not helpful to explain that that is the problem, so putting in this fix might be helpful to some.

**Comment by danmack2** [comment]
Hello I am trying to use mcohen01's procedure for updating home-brew and am have an internal problem get fortran to load.   If we wait for the full package to be posted, will that also require an update to home-brew fortran?

**Comment by offerm** [comment]
Hi Joshua,

Thanks for your great work!

You may want to check this 

> getSymbols("^VIX",auto.assign = F)[as.Date("2016-06-29"),]

It is giving NAs now.
Yahoo sites shows it as zeros.
I don't know how it was before the change but my R code now breaks with the NA so I guess there is some kind of change in behaviour that you may want to check.

Thanks

Offer

**Comment by selgamal** [comment]
@joshuaulrich  Thanks a million for all your work, can't say enough.

I am using the fix, and it is working fine so far in getting the data.

On the other hand, the historical prices in Yahoo have changed, and I just want to confirm my understanding of how the data changed:

It seems to me that now, instead of providing all raw historical prices, the price history is adjusted for splits only, EXCEPT for the "Close" column, it is left as raw price in the csv.

So if I need the raw prices, I will have to unadjust the Open, High and Low columns for splits.

I compared price history of MSFT  as it was provided previously (from a database I been accumulating the past couple of years) and csv downloaded today from yahoo.  I attached a spreadsheet with the comparison.

[MSFT.xlsx](https://github.com/joshuaulrich/quantmod/files/1014177/MSFT.xlsx)

**Comment by Hagakure0c** [comment]
One option is to to use quandl.com for free data or something more expensive like tickdata.com. Yahoo Finance is being treated a bit like Google Finance, not that well.

**Comment by flemingcra** [comment]
I posted a quick fix 2017-05-18 at [http://stackoverflow.com/questions/44015838/cant-download-data-from-yahoo-finance-using-quantmod-in-r/44055578#44055578](url)

**Comment by joshuaulrich** [comment]
@bojanbis Yes, that's the broken version. See [this comment](https://github.com/joshuaulrich/quantmod/issues/157#issuecomment-302159301) for how to install the patched version.

@bdfden @lompoc42 See #152 for a discussion of the issue you encountered.

@herrold Thanks for the information. I'm planning on including a warning to let people know that they should begin to look for alternate data providers.

@rogerjbos Thanks for the note. I noticed that as well, and will add a warning to prevent sending an erroneous date in the request.

@offerm Thanks for the report! I will investigate.

@selgamal Thanks for trying to help! And, yes, I already attempted your suggested fix. ;)

@Hagakure0c Thanks for the note about issues with `runCov()`. The running/rolling functions in TTR cannot handle missing values. Those are called inhomogeneous indicators, and they aren't simple to implement. While one single missing value may be easy to handle, missing values are hard to handle in these functions in a robust and general way).

**Comment by leekaishing** [comment]
Hi after installing the patch from devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
I found the following error message:
Error in .getHandle() : could not find function "get0"
Could you advice what could be the reason behind ? Thanks for the help.

**Comment by lowcsharp** [comment]
Thanks to Joshua Ulrich for the timely response to Yahoo's sudden disconnect. I was able to implement his getSymbols fix and obtain price data. However, there are errors in historical data at the new location. In my case, symbol WBA has incorrect Open and Low entries (i.e: $1.00) on these dates: 
in 2012 - 5/29, 7/5, 9/4, 10/31, 11/23, 12/26
in 2013 - 1/2, 1/22, 2/19, 4/1, 5/28, 7/5
We conclude that this data base is unreliable, especially for back testing purposes. and plan to investigate another source.

**Comment by joshuaulrich** [comment]
@leekaishing `get0` was introduced in R-3.2.0, so you need to be running a version of R that was released in the past 2 years.

**Comment by mlb92** [comment]
@joshuaulrich I have ran your updates

remotes::install_github("joshuaulrich/quantmod")
# or
devtools::install_github("joshuaulrich/quantmod")

and I am still getting the 502 Bad Gateway issue...is there anything you recommend (with an example) for how I can get it to work. Thank you.

**Comment by joshuaulrich** [comment]
@mlb92 You have not run the commands to install the patch.  The commands you used will only work once the patch is merged, which it has not been.  You need to run these commands:

```r
remotes::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
# or
devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")
```

**Comment by mlb92** [comment]
@joshuaulrich I am very new to programming, could you explain what you mean by "You have not run the commands to install the patch."? I have run these lines so far. Thank you for your help:

install.packages("remotes") 
install.packages("devtools")
library(remotes)
library(devtools)

After running these, I get the following errors

> remotes::install_github("joshuaulrich/quantmod") 
Downloading GitHub repo joshuaulrich/quantmod@master
* installing *source* package ‘quantmod’ ...
** libs
gfortran-4.8   -fPIC  -g -O2  -c ohlcq.f -o ohlcq.o
make: gfortran-4.8: No such file or directory
make: *** [ohlcq.o] Error 1
ERROR: compilation failed for package ‘quantmod’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
Warning message:
In utils::install.packages(...) :
  installation of package ‘/var/folders/41/p78vz8291wx4f9k6gpb2c_1c0000gn/T//RtmpK1HTpO/remotesd8e77ec9991/joshuaulrich-quantmod-9305034’ had non-zero exit status
> devtools::install_github("joshuaulrich/quantmod")
Downloading GitHub repo joshuaulrich/quantmod@master
from URL https://api.github.com/repos/joshuaulrich/quantmod/zipball/master
Installing quantmod
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ --no-save --no-restore  \
  --quiet CMD INSTALL  \
  '/private/var/folders/41/p78vz8291wx4f9k6gpb2c_1c0000gn/T/RtmpK1HTpO/devtoolsd8e1f379e94/joshuaulrich-quantmod-9305034'  \
  --library='/Library/Frameworks/R.framework/Versions/3.3/Resources/library' --install-tests 

* installing *source* package ‘quantmod’ ...
** libs
gfortran-4.8   -fPIC  -g -O2  -c ohlcq.f -o ohlcq.o
make: gfortran-4.8: No such file or directory
make: *** [ohlcq.o] Error 1
ERROR: compilation failed for package ‘quantmod’
* removing ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
* restoring previous ‘/Library/Frameworks/R.framework/Versions/3.3/Resources/library/quantmod’
Installation failed: Command failed (1)

**Comment by joshuaulrich** [comment]
@mlb92 Then you should just wait until this is fixed.  I told you to run `remotes::install_github("joshuaulrich/quantmod", ref="157_yahoo_502")` and you ran `remotes::install_github("joshuaulrich/quantmod")`.  You also need to make sure you have the necessary tools needed to install packages with compiled code on Mac (which I have mentioned in a few other comments in this thread).

**Comment by umeshnathani** [comment]
I ran devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502") and ran getSymbols("AAPL"). I get the follwoing:


 Error in download.file(paste(yahoo.URL, "s=", Symbols.name, "&a=", from.m,  : 
  cannot open URL 'https://ichart.finance.yahoo.com/table.csv?s=AAPL&a=0&b=01&c=2007&d=4&e=24&f=2017&g=d&q=q&y=0&z=AAPL&x=.csv'
In addition: Warning message:
In download.file(paste(yahoo.URL, "s=", Symbols.name, "&a=", from.m,  :
  cannot open URL 'https://ichart.finance.yahoo.com/table.csv?s=AAPL&a=0&b=01&c=2007&d=4&e=24&f=2017&g=d&q=q&y=0&z=AAPL&x=.csv': HTTP status was '502 Connection refused'

I am using windows and latest version - R 3.4.0 (also got the same error previously with R 3.3.2)

**Comment by selgamal** [comment]
@umeshnathani  I think you might need to restart your R session after installing the branch (157_yahoo_502) or just remove quantmod package you already have and then run:
devtools::install_github("joshuaulrich/quantmod", ref="157_yahoo_502") and ran getSymbols("AAPL")

I just tested APPL and it worked fine, here are the results and session info:
```
>  library(quantmod)

> getSymbols("AAPL")

[1] "AAPL"

> head(AAPL)

           AAPL.Open AAPL.High AAPL.Low AAPL.Close AAPL.Volume AAPL.Adjusted
2007-01-03     86.29     86.58    81.90   83.80000   309579900      11.97143
2007-01-04     84.05     85.95    83.82   85.66000   211815100      12.23714
2007-01-05     85.77     86.20    84.40   85.04999   208685400      12.15000
2007-01-08     85.96     86.53    85.28   85.47000   199276700      12.21000
2007-01-09     86.45     92.98    85.15   92.57000   837324600      13.22429
2007-01-10     94.75     97.80    93.45   97.00000   738220000      13.85714
> sessionInfo()
R version 3.4.0 (2017-04-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] quantmod_0.4-8 TTR_0.23-1     xts_0.9-7      zoo_1.8-0     

loaded via a namespace (and not attached):
[1] compiler_3.4.0  tools_3.4.0     curl_2.6        grid_3.4.0      lattice_0.20-35
```

**Comment by walts** [comment]
@joshuaulrich The current build of getSymbols.yahoo on github is removing downloaded rows which have a date but only NA prices and volumes.  An example is VDIGX for 2016-06-29.  This seems to be a change from last week's version which didn't remove these rows.  I  had been using na.approx(..., maxgap = 2) to deal with a small number of rows missing in the download and that worked for my purposes.   Could removing missing rows be made an option with a warning if rows were missing rather than the fixed behavior?

**Comment by joshuaulrich** [comment]
@walts That's a fair point.  I made the change to address @ghost's [comment](https://github.com/joshuaulrich/quantmod/issues/157#issuecomment-302697016). While you could use `xts::na.restore()` to put the missing values back, I think it makes more sense to return them with a warning and let the user deal with them.

**Comment by joshuaulrich** [comment]
The `157_yahoo_502` branch is now merged to master.  There may still be some edge-case weirdness, but the main issues appear to be resolved, so I'm closing this as fixed.

Thanks for everyone's feedback, testing, and support!

**Comment by Lois47ed** [comment]
Hi Mr Joshualrich,  I´m Lois.

Since past may 12 of 2017, WannaCry event, I can´t  see this chart (Nike Inc. stok)  in my browser:

http://ichart.finance.yahoo.com/z?s=NKE&t=6m&q=c&l=on&z=m&p=e50,e20&a=v,r14

Dear Joshualrich, ¿Is there any way to fix this problem?

You said that was a change of  URL, this is,   ¨https¨   instead of   ¨http¨

Thanks for your time.

Bye for now

**Comment by Harrymon12** [comment]
Yahoo finance API is not available anymore. I have moved to MarketXLS after this change, much more reliable data.

**Comment by joshuaulrich** [comment]
@Harrymon12 It simply isn't true that the Yahoo Finance API is not available.  I just ran this code:

```r
R> quantmod::getSymbols("SPY", src = "yahoo")
[1] "SPY"
R> head(SPY)
           SPY.Open SPY.High SPY.Low SPY.Close SPY.Volume SPY.Adjusted
2007-01-03  176.779  177.537 174.691    141.37   94807600     113.7571
2007-01-04  175.512  176.531 174.741    141.67   69620600     113.9985
2007-01-05  175.636  175.723 174.455    140.54   76645300     113.0893
2007-01-08  175.002  175.735 174.294    141.19   71655000     113.6122
2007-01-09  175.611  175.971 174.480    141.07   75680100     113.5157
2007-01-10  174.704  175.934 174.356    141.54   72428000     113.8939
R> packageVersion("quantmod")
[1] '0.4.10'
```

**Comment by Harrymon12** [comment]
@joshuaulrich Oh I see. It came back.

**Comment by chadworth** [comment]
It looks like same underlying data source is used for both the query1 and query2 APIs. E.g. the WBA issue that @lowcsharp mentioned (https://github.com/joshuaulrich/quantmod/issues/157#issuecomment-303390934) is an issue in both endpoints
